[libc][math][c23] Add rsqrtf16() function

amemov opened the pull request:
Addresses #132818 
Part of #95250

amemov committed:
- rsqrtf16 refactored

amemov committed:
Clang-formated the files

amemov committed:
Replaced the computation for valid X with polynomial approximation

amemov committed:
Added range reduction to the approximation

amemov commented:
Trying to figure out what would be the best option to compute the result.
I found that the current polynomial produces the least errors ( bigger ones yield negligible results )
`P = fpminimax(1/sqrt(x), [|0,1,2,3,4,5|], [|SG...|], [0.5, 1]);
`
And has ULP Error = 1.0



Also found this already existing implementation: https://github.com/llvm/llvm-project/blob/ae6b4b23ea4291e937192a3c08d0f3c9835864c2/libc/src/__support/fixed_point/sqrt.h#L39
It has some other interesting points that I found when I was doing my research: specifically, Newton's method. I tried adding one iteration of it, but same: it didn't help much.

amemov committed:
Added Newton-Raphson iterations
-The accuracy improved drastically, but it still fails

amemov committed:
Added separate handling for mantissa == 0.5f. Resulted in fewer errors

amemov committed:
- Fixed ULP errors
- Refactored the implementation to match the proposal for constexpr
- Added rsqrtf16 in Bazel build

amemov committed:
clang-formatted the files

amemov commented:
@overmighty @lntue

amemov committed:
Formatted BUILD.Bazel w/ buildifier

lntue commented on libc/src/__support/math/rsqrtf16.h:
```diff
@@ -0,0 +1,139 @@
+
+  // x = +inf => rsqrt(x) = 0
+  if (LIBC_UNLIKELY(xbits.is_inf())) {
+    return fputil::cast<float16>(0.0f);
```
`FPBits::zero().get_val()`?

lntue commented:
> > > Trying to figure out what would be the best option to compute the result. I found that the current polynomial produces the least errors ( bigger ones yield negligible results ) `P = fpminimax(1/sqrt(x), [|0,1,2,3,4,5|], [|SG...|], [0.5, 1]); ` And has ULP Error = 1.0
> > > Also found this already existing implementation:
> > > https://github.com/llvm/llvm-project/blob/ae6b4b23ea4291e937192a3c08d0f3c9835864c2/libc/src/__support/fixed_point/sqrt.h#L39
> > > 
> > > It has some other interesting points that I found when I was doing my research: specifically, Newton's method.
> > > Upd: Tried adding 2 iterations of Newton's method. Each significantly reduced number of errors, but there are still some
> > 
> > 
> > Can you compare the performance of this with
> > ```
> >   fputil::cast<float16>(1.0f / fputil::sqrt(fputil::cast<float>(x)));
> > ```
> 
> I wrote this test to check the performance of the implementation and ran the tests for rsqrtf16 a few times:
> 
> ```
> TEST_F(LlvmLibcRsqrtf16Test, PositiveRange_OneOverSqrtFputil) {
>   for (uint16_t v = POS_START; v <= POS_STOP; ++v) {
>     float16 x = FPBits(v).get_val();
> 
>     float16 y = LIBC_NAMESPACE::fputil::cast<float16, float>(
>         1.0f / LIBC_NAMESPACE::fputil::sqrt<float, float>(
>                    LIBC_NAMESPACE::fputil::cast<float, float16>(x)));
> 
>     EXPECT_MPFR_MATCH_ALL_ROUNDING(mpfr::Operation::Rsqrt, x, y, 1.0);
>   }
> }
> ```
> 
> Turns out that my implementation is ~3x slower than just directly calling `1.0f / fputil::sqrt` :/ Not sure why is that - because of too many branches or I wrote over-complicated approximation. The one you see is the most minimal I was able to derive so far - I started with 7-degree polynomial and 2 iterations of Newton's method and was able to reduce it to 5-degree and 1 iteration. What do you think?

It is actually expected, because single/double precision division and square root in modern hardware are quite efficient.
You can see for example Zen3 in https://www.agner.org/optimize/instruction_tables.pdf
SQRTSS and DIVSS latencies are 14 and 10.5 clocks respectively, while ADDSS/MULPS and VFMA are 3 and 4 clocks.

So unless you can reduce to maybe 3, 4 multiply-adds, the extra logic like branching, exponent reductions, ... around the computations will make it slower than the straightforward sqrt + div in single precision.

For rsqrtf16, the newton-raphson method will be better than sqrt + div for targets without single precision hardware, such as some embedded system.  But in that case, you will need to implement polynomial approximation + newton raphson in integer / fixed point arithmetic to gain the efficiency.

amemov committed:
- Refactored the rsqrtf16 implementation to temporarily use hardware for calling sqrt
- Adjusted the results from fixed-point call of hardware instruction to match the correctness required in the Libc
- TODO: In the next PR add int-based approximation of the function for scenario where floats are not available in the hardware

lntue commented on libc/src/__support/math/rsqrtf16.h:
```diff
@@ -0,0 +1,84 @@
+namespace LIBC_NAMESPACE_DECL {
+namespace math {
+
+static constexpr float16 rsqrtf16(float16 x) {
```
`LIBC_INLINE`

lntue commented on libc/src/__support/math/rsqrtf16.h:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  // x = +inf => rsqrt(x) = 0
+  if (LIBC_UNLIKELY(xbits.is_inf())) {
```
no `{ }` for a single statement.

lntue commented on libc/src/__support/math/rsqrtf16.h:
```diff
@@ -0,0 +1,84 @@
+  // Targeted post-corrections to ensure correct rounding in half for specific
+  // mantissa patterns
+  const uint16_t half_mantissa = x_abs & 0x3ff;
+  if (half_mantissa == 0x011F) {
```
`LIBC_UNLIKELY` on both exceptional cases.

lntue commented on libc/config/linux/x86_64/entrypoints.txt:
```diff
@@ -784,6 +784,7 @@ if(LIBC_TYPES_HAS_FLOAT16)
     libc.src.math.rintf16
     libc.src.math.roundevenf16
     libc.src.math.roundf16
+    libc.src.math.rsqrtf16
```
also add entrypoints for other targets that have float16.

amemov commented:
For the record - the implementation had to be changed because it wasn't as fast as calling the hardware specific instructions by a very huge margin. Below are the observations that I saw when I was comparing 4 different implementations using Google Benchmark and [suite](https://blog.llvm.org/posts/2025-08-29-gsoc-profiling-and-testing-math-functions-on-gpus/) introduced as part of GSoC 2025:
```
Run on (14 X 4900 MHz CPU s)
CPU Caches:
L1 Data 48 KiB (x7)
L1 Instruction 64 KiB (x7)
L2 Unified 2048 KiB (x7)
L3 Unified 12288 KiB (x1)
Load Average: 1.12, 1.53, 1.24
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-- LOG(0): Value passed to --benchmark_min_time should have a suffix. Eg., `30s` for 30-seconds.-- LOG(0): Value passed to --benchmark_min_time should have a suffix. Eg., `30s` for 30-seconds.-- LOG(0): Value passed to --benchmark_min_time should have a suffix. Eg., `30s` for 30-seconds.-- LOG(0): Value passed to --benchmark_min_time should have a suffix. Eg., `30s` for 30-seconds.---------------------------------------------------------------------
Benchmark           Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------
BM_Current     632704 ns       629457 ns         2220 items_per_second=50.4324M/s
BM_Impl1      1058891 ns      1053017 ns         1339 items_per_second=30.1467M/s
BM_Impl2       686397 ns       684024 ns         2010 items_per_second=46.4092M/s
BM_ViaSqrt     402577 ns       401175 ns         3345 items_per_second=79.13M/s
```
Where `Current` is what I wrote locally (didn't push on PR), `Impl2` is similar to current but has slightly bigger polynomial, `Impl1` is what you see on Github in the previous commit, and `ViaSqrt` is `LIBC_NAMESPACE::fputil::cast<float16>(1.0f / LIBC_NAMESPACE::fputil::sqrt<float>(LIBC_NAMESPACE::fputil::cast<float>(x)));`

However, the `ViaSqrt` as described above was not able to satisfy the correctness requirements established in the Libc across 2 mantissas: 0x0313 and 0x011F. Therefore, I added a small correction at the end. With this changes the implementation is a little slower than directly calling `ViaSqrt` (by ~5000 ns), but still way faster than the best implementation I did ( which is `Current` ) by 200000 ns.

With that being said, there is still some work left for the future: if hardware provides sqrt() instruction - it should be used, otherwise an int-based math approximation should be used for targets that don't have `LIBC_TARGET_CPU_HAS_FPU_FLOAT`

amemov committed:
- Addressed the comments and added entrypoints to other targets in Linux

lntue commented on libc/test/src/math/smoke/rsqrtf16_test.cpp:
```diff
@@ -0,0 +1,37 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/__support/libc_errno.h"
```
without `libc_errno = 0` below, change this to `hrd/errno_macros.h` and update the dependency accordingly.

amemov committed:
- Replaced src/__support/libc_errno.h with hdr/errno_macros.h
